File: operation.py
import math
import abc


class Operation(abc.ABC):
    @abc.abstractstaticmethod
    def operate(a, b):
        pass

    @abc.abstractstaticmethod
    def to_string(a, b):
        pass


class Add(Operation):
    def operate(a, b):
        return a + b

    def to_string(a, b):
        return str(a) + " + " + str(b)


class Subtract(Operation):
    def operate(a, b):
        return a - b

    def to_string(a, b):
        return str(a) + " - " + str(b)


class Multiply(Operation):
    def operate(a, b):
        return a * b

    def to_string(a, b):
        return str(a) + " * " + str(b)


class Divide(Operation):
    def operate(a, b):
        return a / b

    def to_string(a, b):
        return str(a) + " / " + str(b)


class Exponent(Operation):
    def operate(a, b):
        # for some reason, python detects numeric overflow floats, but not ints
        return a ** float(b)

    def to_string(a, b):
        return str(a) + " ^ " + str(b)


class Log(Operation):
    def operate(a, b):
        return math.log(a, b)

    def to_string(a, b):
        return "log_" + str(a) + "(" + str(b) + ")"

File: src/operation.ts
export interface Operation {
  operate(a: number, b: number): number;
  toString(a: number | string, b: number | string): string;
}

export const Add: Operation = {
  operate(a: number, b: number): number {
    return a + b;
  },

  toString(a: number | string, b: number | string): string {
    return `${a} + ${b}`;
  },
};

export const Subtract: Operation = {
  operate(a: number, b: number): number {
    return a - b;
  },

  toString(a: number | string, b: number | string): string {
    return `${a} - ${b}`;
  },
};

export const Multiply: Operation = {
  operate(a: number, b: number): number {
    return a * b;
  },

  toString(a: number | string, b: number | string): string {
    return `${a} * ${b}`;
  },
};

export const Divide: Operation = {
  operate(a: number, b: number): number {
    return a / b;
  },

  toString(a: number | string, b: number | string): string {
    return `${a} / ${b}`;
  },
};

export const Exponent: Operation = {
  operate(a: number, b: number): number {
    return a ** b;
  },

  toString(a: number | string, b: number | string): string {
    return `${a} ^ ${b}`;
  },
};

export const Log: Operation = {
  operate(a: number, b: number): number {
    return Math.log(b) / Math.log(a);
  },

  toString(a: number | string, b: number | string): string {
    return `log_${a}(${b})`;
  },
};
